### ISSUE TITLE
GIF search

### ISSUE THREAD

**Issue by aircave** [florisboard/florisboard#53]
Any chance gif searching would ever be added? Not sure of any other keyboards that have it other than Gboard.

**Comment by D3SOX** [comment]
I think there is a Giphy API and an SDK. For privacy reasons I would implement it using the API, similar to Signal (which is also open source) and it should be no problem.

**Comment by kojid0** [comment]
So florisboard has internet access permissions? 🤔

**Comment by RickyM7** [comment]
> So florisboard has internet access permissions? 🤔

@kojid0 No, and I'd really like it to stay that way, but maybe this feature can be added as dictionaries are added on some keyboards, i.e. adding this feature as an extra app. I don't know if this would really solve the necessary internet access problem, but it would prevent the app size from increasing too much and leave this feature as something additional for anyone who really wants to use (I personally wouldn't use this feature).

**Comment by kojid0** [comment]
> > So florisboard has internet access permissions? 🤔
> 
> @kojid0 No, and I'd really like it to stay that way, but maybe this feature can be added as dictionaries are added on some keyboards, i.e. adding this feature as an extra app. I don't know if this would really solve the necessary internet access problem, but it would prevent the app size from increasing too much and leave this feature as something additional for anyone who really wants to use (I personally wouldn't use this feature).

I wouldn't use this feature either.
Yeah, a separate "Florisboard Plus" app with GIF support (maybe free on F-Droid and Paid on Google Play to support the dev?) would be a solution. Or, the other way around, "Florisboard Slim" with only basic functions/permissions.

**Comment by D3SOX** [comment]
I don't see why this would significantly increase the app size. It's basically just a view with a search box that queries the API and displays the images. I think a setting to enable/disable this is appropriate.

**Comment by RickyM7** [comment]
> I wouldn't use this feature either.
Yeah, a separate "Florisboard Plus" app with GIF support (maybe free on F-Droid and Paid on Google Play to support the dev?) would be a solution. Or, the other way around, "Florisboard Slim" with only basic functions/permissions.

@kojid0 I did not mean to refer to a separate version of FlorisBoard itself, but rather just a small separate app with only the Gifs function, however this is also an option, I think.

**Comment by flide** [comment]
Hi All, 
I am the author of 8vim keyboard for android. I also had a similar problem of wanting to integrate emojis, lennys and gifs to my keyboard. I initially tried to cram all into the keyboard, but it did not feel right. The app size kept increasing and I figured that the user will end up missing out on features that I am not even thinking about. 
So, what I did was to cut off the emoji and lenny boards into it's seperate keyboard 'pictograph' and provided an option in the settings to allow the user to configure a keyboard they want to open when they want to input emoji's and other similar stuff.

Although, I still love the idea, most other emoji keyboards don't provide an easy way back to the original keyboard, so ... that's an issue. 

If you want you can try that route. A seperate internet connected keyboard does not effect your keyboard's privacy and allows the focus to remain on text input. 

If interested, you can check for the relavent code in my 8vim repo.

**Comment by flide** [comment]
As a side note, folks here at florisboard can try using the pictograph board as a testing ground and maybe help to introduce gif's to it. It already has emoji's (3 variations) and lennys which I plan to make configurable.

**Comment by RickyM7** [comment]
@flide First congratulations on your project, it is very interesting.
I just tested 8Vim with Pictograph and the integration between them occurs in a very fluid way, in addition the pictograph is quite complete in relation to the midias (emoji, kaomojis, etc).
Pictograph is a completely separate app from 8Vim, which makes you have the alternative of having it together with 8Vim or not, this also allows the app to have functions that can be considered non-privacy friendly without worries by being separated from the original app (just like you said).
I think it would be wonderful if FlorisBoard did something like this for the complementary features, e.g. Emojis (custom versions only, the option of emojis that FlorisBoard currently has does not consume much space), Gifs, Dictionaries, etc.

Note: @flide I think the emoji button in 8Vim should be hidden when no emoji app is selected.

**Comment by patrickgold** [comment]
First of all, thanks all for your great input to this issue. You gave me some good ideas for this rather difficult-to-implement-while-being-privacy-firendly topic.

Putting privacy things aside, it isn't too difficult to add a GIF feature to FlorisBoard via the Giphy API. As @D3SOX correctly said, it wouldn't even increase the app size that much, as only the media UI layout has to adjusted and some code has to be added to interact with Giphy and fetch the GIFs (I wouldn't use their SDK, as it is a blackbox and I don't know what it really does).

Now coming to the privacy side: Implementing GIFs require the network permission, which FlorisBoard currently doesn't have. And I want it to stay like this. And to be honest, as much as I would like to implement GIFs into FlorisBoard, I do not want to sacrifice the no-network-access philosophy for this feature.

I really like the idea of @RickyM7 though, to make the GIF functionality a separate extension app. Only this extension has the internet permission, which then fetches the GIFs and shares it with FlorisBoard.

Note, that I will soon begin working on v0.4.0 and composing suggestions though, which is a ton of work, so if I start working on GIFs, it won't be until after the v0.4.0 release. Coming with composing suggestions, I also plan to make dictionaries as extensions, which could then also serve as a base for the GIFs and make implementing such much easier.

@flide Thanks for checking in on FlorisBoard. Your concept of splitting the text and media into two separate IMEs is also a great way to tackle this issue. I will surely check out your implementation if I choose to go your way!

**Comment by aircave** [comment]
@patrickgold 
Thank you for the consideration and thanks to everyone working on this project.
The GIF extension/addon all while retaining privacy would appeal to many, IMHO.

**Comment by singpolyma** [comment]
Giphy even supports animated webp via API, which would be ideal in terms of quality / support tradeoff

**Comment by VioletStorm** [comment]
If this feature is added I'd like to suggest also including a toggle to deactivate the content filter, this is presently only achievable on gboard with root (and it only works so-so). It would be quite wonderful to be able to disable result censorship natively with a simple toggle

**Comment by ItsRogueRen** [comment]
Gifs would 100% make me swap from Openboard. That team has refused to add gifs and I use them daily in texts with friends.

**Comment by Robouste** [comment]
Why is this a low priority? That seems like a must have to me. I would LOVE to try this keyboard but it needs to support gifs.

**Comment by lm41** [comment]
> Why is this a low priority?

Because it isn't easy to implement gifs in a privacy friendly way.

**Comment by VioletStorm** [comment]
I agree that this ought not to be a low priority feature, it isn't critical but it's important!

I don't care about perfect privacy for gif search, all I care about is that the results are unfiltered as gboard currently does. I'd imagine it wouldn't be that hard though? There are a number of private and unfiltered search APIs out there! Does duckduckgo have an API?

**Comment by VioletStorm** [comment]
> Giphy even supports animated webp via API, which would be ideal in terms of quality / support tradeoff

Unfortunately giphy is very heavily filtered I believe, I wouldn't want anything but unfiltered or at least optionally unfiltered

**Comment by gifboard** [comment]
@VioletStorm As an alternative, since I couldn't find the feature anywhere apart from gboard, I created https://github.com/gifboard/gifboard which is a standalone GIF-only keyboard (you use it alongside a regular keyboard and switch between input methods)

It does rely on google to search for gifs, but includes Advanced Settings to set the Content Filter (safesearch) parameters

**Comment by VioletStorm** [comment]
> [@VioletStorm](https://github.com/VioletStorm) As an alternative, since I couldn't find the feature anywhere apart from gboard, I created https://github.com/gifboard/gifboard which is a standalone GIF-only keyboard (you use it alongside a regular keyboard and switch between input methods)
> 
> It does rely on google to search for gifs, but includes Advanced Settings to set the Content Filter (safesearch) parameters

Oh my goodness, this is absolutely perfect!! I can't believe you just up and made this, but I'm exceedingly grateful!!!

**Comment by patrickgold** [comment]
@gifboard Is there a reason your keyboard only has one commit and your history was force-pushed today?

Not to be mean here, I want to encourage open-source, but your account is 3 hours old and you just drop a full codebase without any commit history, no publications on app stores, something feels phishy to me.

**Comment by gifboard** [comment]
@patrickgold I gotta start somewhere, I've been waiting for an alternative for a while and nobody ever made one so I decided to make one for myself. After testing/iterating on it internally for a few weeks, I figured others might want to use it too, so I just made it open source and am sharing it now for the first time.

Using a new separate account since I don't want future employers to see a potentially NSFW app linked to my developer portfolio. I squash all commits and past releases since I just released it now, and there's no need for the commit history being public and showing a hundred non-working intermediate commits. I can't publish it on app stores for obvious reasons, publishing on F-Droid is in progress.

Feel free to inspect the source (it's a pretty simple app) and build it yourself, I've provided step-by-step instructions and a Makefile for easy builds too. Feel free to open any discussions or issues in the repo itself so as to not go off-topic on this thread.

**Comment by patrickgold** [comment]
Absolutely, you gotta start somewhere and your reasoning makes sense to me, sorry if I came off as unfriendly.

**Comment by PumaFPV** [comment]
Gotta confirm, gifboard is a great solution to the problem I was trying to fix. (Get access to gifs without using gboard)

**Comment by VioletStorm** [comment]
Oh yeah nah I use it daily now xD it's so good
  

  

  

  
  
  
  
  
>

**Comment by Robouste** [comment]
Great work, I'm impressed by the quality and the nice tutorial. Thank you

**Comment by gifboard** [comment]
Thank you. I'm looking for help on https://github.com/gifboard/gifboard/issues/4 to find a way to integrate GifBoard as a 3rd party GIF plugin for FlorisBoard in a clean manner without violating the offline constraint on FlorisBoard. I tried a few different approaches which I documented in the issue, but the implementation appears to be beyond my abilities for now. If anyone is willing to try it out, I would appreciate ideas and contributions.

**Comment by VioletStorm** [comment]
I took would love this
  

  

  

  
  
  
  
  
>

**Comment by mozzwald** [comment]
I've submitted a PR to Heliboard with "media plugin" support. It allows the keyboard to present a media display picker window, but it gets all media via android IPC from another app/plugin. I made two demo plugins; GIF search and Webdav browse. This keeps the keyboard free of any new (network) permissions and restricts those permissions to the plugins for users who do not need/want GIFs. It also keeps you in the same keyboard without having to switch to another IME. In the PR I was asked if any other keyboards had something similar so I went searching and found Florisboard which has limited addon/extension support. I figured this idea might also work here. 

https://github.com/HeliBorg/HeliBoard/pull/2500

**Comment by VioletStorm** [comment]
We definitely need it   
  

  

  

  
  
  
  
  
>
